### DISCUSSION TITLE
Ideas

### DISCUSSION THREAD

**Discussion by oberblastmeister** [tjdevries/rofl.nvim#7] [Ideas]
What I am thinking for this plugin is to have the fastest and smoothest completion. I think we should focus on that first and then do stuff like the lsp hovery stuff. My idea is to have some builtin sources in rust like buffers and paths for speed and then have some programmable lua sources.

**Comment by tjdevries** [comment]
Yes, I agree that we can have some fun and interesting sources built in to `rofl` in Rust.

Before we write too much code for doing this though, I need to spend some time thinking about the interfaces and what Neovim expects on it's side.

Additionally, we can also think of new APIs that could be helpful in Neovim core, as I'm sure other users have ideas for things they would like (such as Shougo, or others making completion plugins). We can always submit patches upstream to add new features :)

**Reply by OldManFroggy** [reply]
I would definitely 2nd on basic completion before the fun glossy stuff, esp. as a RUST dev moving to NVIM for the first time. I have found a lot of completion tools seem to be focused on the gloss while ignoring the basics - and yes rust_analyzer gives a lot and people want to show it off, but no need to rush.

just my 2c...

**Comment by tjdevries** [comment]
In regards to lua sources, I 100% agree. I want to be able to configure, run and implement new sources in Lua (and other languages) so it's super easy to make new sources for people.

**Reply by oberblastmeister** [reply]
I have an idea for how to implement lua sources. So we have a private table in the rofl lua file, and we have a function defined on that table. That function will add a user-defined function and the user also provides a name. It will notify the binary of the name of the function. Then on the rust side all we have to do is `luaeval` by indexing into the table and calling the function and then we can get the entries of the source. Then we can just concurrently merge the results with the other sources.
